### ISSUE TITLE
steep misunderstands string slice types

### ISSUE THREAD

**Issue by p-datadog** [soutaro/steep#1219]
Correct me if I'm wrong but I don't think a string slice that begins at the beginning of a string can ever be nil.

For example steep complains of the following code:

```
a=''
a[0..0]+'x'
```

Thusly:

```
test.rb:2:7: [error] Type `(::String | nil)` does not have method `+`
│ Diagnostic ID: Ruby::NoMethod
│
└ a[0..0]+'x'
         ~

```

`nil` is not a possible type for `a[0..0]` when `a` is a String.

**Comment by soutaro** [comment]
Yeah, it cannot be `nil`, but Steep cannot identify that the range starts with `0`.
Similar to `array.find {}` always returns `nil` if it's empty, but Steep cannot find it.

Those properties around values cannot be written as types, and Steep cannot do much for them.
(There are much more complicated type systems called *dependent types* to handle these problems, but Steep/RBS is not something like it.)

**Comment by p-datadog** [comment]
OK, is there any way to not have steep errors in these cases of valid Ruby code?
